# Practice on STRING_AGG() & ARRAY_AGG()
We will use the Google Analytics dataset `data-to-insights.ecommerce.all_sessions_raw`.

In [1]:
# This cell is to enable the "hint" functionality. After each question there is a cell with either a hint about the correct answer or the solution.
from IPython.display import Pretty as disp
hint = 'https://raw.githubusercontent.com/soltaniehha/Business-Analytics-Toolbox/master/docs/hints/'  # path to hints on GitHub

Question: Find out how many product names and product SKUs are on the website?

In [3]:
%%bigquery --project=ba-770
SELECT COUNT(*) FROM
(
  SELECT DISTINCT productSKU, v2ProductName
  FROM `data-to-insights.ecommerce.all_sessions_raw`
)

Query is running:   0%|          |

Downloading:   0%|          |

f0_
0  2273

Now find the number of distinct SKUs:

In [5]:
%%bigquery --project=ba-770
SELECT COUNT(DISTINCT productSKU)
FROM `data-to-insights.ecommerce.all_sessions_raw`

Query is running:   0%|          |

Downloading:   0%|          |

f0_
0  1909

Obviously these numbers do not match which indicates that there are duplications. Let's determine which products have more than one SKU and which SKUs have more than one Product Name.

Let's determine if some product names have more than one SKU:

In [6]:
%%bigquery --project=ba-770
SELECT
  v2ProductName,
  COUNT(DISTINCT productSKU) AS SKU_count,
  STRING_AGG(DISTINCT productSKU LIMIT 5) AS SKU
FROM `data-to-insights.ecommerce.all_sessions_raw`
WHERE productSKU IS NOT NULL
GROUP BY v2ProductName
HAVING SKU_count > 1
ORDER BY SKU_count DESC

Query is running:   0%|          |

Downloading:   0%|          |

v2ProductName  SKU_count  \
0             Waze Women's Typography Short Sleeve Tee         12   
1                                    Google Sunglasses         10   
2          Google Men's Watershed Full Zip Hoodie Grey         10   
3    Android Women's Short Sleeve Badge Tee Dark He...         10   
4           Google Women's Insulated Thermal Vest Navy         10   
..                                                 ...        ...   
488                            Android 5-Panel Low Cap          2   
489      Google Women's Badge Performance Tee Charcoal          2   
490  Android 24 oz Button Lid Sport Water Bottle Green          2   
491                               Porcelain Mosaic Mug          2   
492                     Google Leather Journal (Black)          2   

                                                   SKU  
0    GGOEWALJ083414,GGOEWALJ083417,9184707,9184709,...  
1    9180829,GGOEGHGC019799,GGOEGHGH019699,GGOEGHGT...  
2    9182739,GGOEGADJ056813,GGOEGADJ056816,GGOEGADJ...  
3    GGOEGAAX0282,GGOEAAEJ028214,9182759,9182179,91...  
4    GGOEGAPL058516,GGOEGAPL058513,GGOEGAPL058514,G...  
..                                                 ...  
488                             GGOEAHPB080210,9184624  
489                               GGOEGAAX0579,9182773  
490                             9180751,GGOEADHH015499  
491                                  10 52048,10 52246  
492                             GGOEGOCB078299,9183221  

[493 rows x 3 columns]

We can see that 493 products are under this category. We can see the SKUs that these product names are related to.

## Your turn
Find the SKUs that have multiple product names:

productSKU  product_count  \
0    GGOEGEVA022399              3   
1    GGOEGCLB020832              3   
2    GGOENEBJ079499              3   
3           9180752              3   
4    GGOEGBMC056599              3   
..              ...            ...   
342         9180775              2   
343         9182647              2   
344         9180872              2   
345  GGOEGAEJ030114              2   
346  GGOEGAEJ028916              2   

                                          product_name  
0    Micro Wireless Earbud,Micro Wireless Earbuds,A...  
1    Softsided Travel Pouch Set,Set of 3 Nested Tra...  
2    Nest® Learning Thermostat 3rd Gen-USA - Stainl...  
3    Android 24 oz Contigo Bottle,Android 24 oz But...  
4    Waterpoof Gear Bag,Waterproof Gear Bag,Google ...  
..                                                 ...  
342             Google Pet Feeding Mat,Pet Feeding Mat  
343  Google Women's Short Sleeve Badge Tee Ice Gray...  
344   Collapsible Pet Bowl,Google Collapsible Pet Bowl  
345  Google Women's Tank Top,Google Women's Tank To...  
346  Google Women's Short Sleeve Hero Dark Grey,Goo...  

[347 rows x 3 columns]

In [8]:
# SOLUTION: Uncomment and execute the cell below to get help
#disp(hint + '06-02-products')